javalib/src/main/scala/java/util/RedBlackTree.scala

```diff
@@ -0,0 +1,1019 @@
+      start: A, startKind: BoundKind, end: A, endKind: BoundKind)(
+      implicit comp: Comparator[_ >: A])
+      extends AbstractTreeIterator[A, B, R](
+          tree,
+          DescendingTreeIterator.nullIfBeforeEnd(
+              maxNodeBefore(tree.root, start, startKind), end, endKind)) {
+
+    protected final def advance(node: Node[A, B]): Node[A, B] =
+      DescendingTreeIterator.nullIfBeforeEnd(predecessor(node), end, endKind)
+  }
+
+  private[this] object DescendingTreeIterator {
+    @inline
+    private def nullIfBeforeEnd[A, B](node: Node[A, B], end: A,
+        endKind: BoundKind)(implicit comp: Comparator[_ >: A]): Node[A, B] = {
+      if (endKind != NoBound && (node ne null) &&
+          !isWithinLowerBound(node.key, end, endKind)) {
+        null
+      } else {
+        node
+      }
+    }
+  }
+
+  private[this] final class DescendingEntriesIterator[A, B](tree: Tree[A, B],
+      start: A, startKind: BoundKind, end: A, endKind: BoundKind)(
+      implicit comp: Comparator[_ >: A])
+      extends DescendingTreeIterator[A, B, Map.Entry[A, B]](tree, start, startKind, end, endKind) {
+
+    def nextResult(node: Node[A, B]): Map.Entry[A, B] = node
+  }
+
+  private[this] final class DescendingKeysIterator[A, B](tree: Tree[A, B],
+      start: A, startKind: BoundKind, end: A, endKind: BoundKind)(
+      implicit comp: Comparator[_ >: A])
+      extends DescendingTreeIterator[A, B, A](tree, start, startKind, end, endKind) {
+
+    def nextResult(node: Node[A, B]): A = node.key
+  }
+
+  private[this] final class DescendingValuesIterator[A, B](tree: Tree[A, B],
+      start: A, startKind: BoundKind, end: A, endKind: BoundKind)(
+      implicit comp: Comparator[_ >: A])
+      extends DescendingTreeIterator[A, B, B](tree, start, startKind, end, endKind) {
+
+    def nextResult(node: Node[A, B]): B = node.value
+  }
+
+  // building
+
+  /** Common implementation of `fromOrderedKeys` and `fromOrderedEntries`. */
+  @noinline
+  def fromOrdered[A, B, C](xs: Iterator[A], size: Int,
+      keyOf: js.Function1[A, B], valueOf: js.Function1[A, C]): Tree[B, C] = {
+    // maximum depth of non-leaf nodes == floor(log2(size))
+    val maxUsedDepth = 32 - Integer.numberOfLeadingZeros(size)
+
+    @noinline
+    def createSubTree(level: Int, size: Int): Node[B, C] = size match {
+      case 0 =>
+        null
+      case 1 =>
+        /* Nodes on the last level must be red, because the last level might
+         * not be full. If they were black, then the paths from the root to the
+         * last level would have 1 more black node than those ending at the
+         * second-to-last level.
+         * Exception: when the only node is the root, because the root must be
+         * black.
+         */
+        val item = xs.next()
+        val red = level == maxUsedDepth && level != 1
+        Node.leaf(keyOf(item), valueOf(item), red, null)
+      case _ =>
+        val leftSize = (size - 1) >> 1 // opt: (size - 1) / 2 with size > 0
```

Comment:
I do not believe this is a reasonable performance optimization. Any halfway decent compiler will do this on its own. (note that the condition `size > 0` is not necessary).

Reply:
For negative numbers on the left, `>>` does not round in the same direction as `/`, so compilers *don't* do that on their own.

It's a rewriting we often do manually in this codebase. There was one recently in `PriorityQueue`.

Reply:
Right... I missed that... (tested on py which of course has different behavior... :()